x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -392,3 +391,102 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
 	}
 	return to
 }
+
+// distributeRewards is the main function for reward distribution for providers and delegators
+func (k Keeper) distributeRewards(ctx sdk.Context, providerAddr sdk.AccAddress, chainID string, block uint64, totalReward math.Int) error {
+	epoch, _, err := k.epochStorageKeeper.GetEpochStartForBlock(ctx, block)
+	if err != nil {
+		return utils.LavaFormatError("could not calculate provider reward by delegations", err,
+			utils.Attribute{Key: "block", Value: block},
+		)
+	}
+	stakeStorage, found := k.epochStorageKeeper.GetStakeStorageEpoch(ctx, epoch, chainID)
+	if !found {
+		return utils.LavaFormatError("could not calculate provider reward by delegations", fmt.Errorf("stake storage for epoch not found"),
+			utils.Attribute{Key: "epoch", Value: epoch},
+			utils.Attribute{Key: "chainID", Value: chainID},
+		)
+	}
+
+	stakeEntry, found, stakeIndex := k.epochStorageKeeper.GetStakeEntryByAddressFromStorage(ctx, stakeStorage, providerAddr)
+	if !found {
+		return utils.LavaFormatError("could not calculate provider reward by delegations", fmt.Errorf("provider stake entry not found"),
+			utils.Attribute{Key: "provider", Value: providerAddr},
+			utils.Attribute{Key: "chainID", Value: chainID},
+			utils.Attribute{Key: "totalRewards", Value: totalReward},
+		)
+	}
+
+	providerReward, err := k.calculateProviderDelegatorsRewards(ctx, stakeEntry, stakeIndex, totalReward)
```

Comment:
as per comment to the function called, this should be:
```
providerReward, delegatorsReward := ....
```

Reply:
outdated. There is no such function.
Also, CalcRewards was implemented here: https://github.com/lavanet/lava/commit/f9c49a16a43c79f77ffae7432d69b349a014f91a